core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -643,7 +645,12 @@ private[spark] object Utils extends Logging {
           fileOverwrite)
       } else {
         val in = fs.open(innerPath)
-        val targetFile = new File(targetDir, innerPath.getName)
+        val targetFile: File =
+          if (!filename.isEmpty) {
+            new File(targetDir, filename)
+          } else {
+            new File(targetDir, innerPath.getName)
+          }
```

Comment:
I think it's a little weird but it works. Only the very first call to `fetchHcfsFile` defines `filename`. If the `path` passed to it is a directory, it will recursively call itself without setting `filename`. If it's a file, it will write the file using the given `filename`. So even if it could be clearer, the code as is should work.

But I'm ok with making it clearer, exactly to avoid this kind of discussion. :-)